Add more tests

Andrew-Bx opened the pull request:
PR comment: restoring a few more automated tests, and adding to the CI workflow

Andrew-Bx reviewed (commented):
N/A

Andrew-Bx commented on test/disabled/id-tests.js:
```diff
@@ -1,76 +0,0 @@
-describe( 'Dynamic ids work properly', function(){
```
looks like the concept of 'dynamic ids' has been removed, so simply deleting these tests

Andrew-Bx commented:
Re acacff4 - the tests imply that it should be possible to start with an empty stack.  To get that to work had to let Stack._activeComponentItem be nullable (and handle that to avoid null-ref errors).  So this now works, but it seems to me a bit inconsistent in that you cannot remove the last item from the stack (ie make it empty again)... so not sure if this is the way we really want to go?

Andrew-Bx commented:
Re 8294e84 - N.B. removed stuff to do with un/binding a 'context'

Andrew-Bx commented:
Re 359840f - changed tasks so that 'test' runs headless, and test:watch runs chrome.  But maybe other way around would be better?  i.e. specify multiple headless browsers in karma.conf (and let these get run by CI), and provide alternative task(s) for running tests against non-headless-browser (ie for running the tests manually ?
(hopefully this is 'good enough for now', but perhaps in future might want to look into using one of:
- https://www.npmjs.com/package/karma-testingbot-launcher
- https://www.npmjs.com/package/karma-browserstack-launcher
- https://www.npmjs.com/package/karma-detect-browsers
)

martin31821 commented on src/ts/config/resolved-config.ts:
```diff
@@ -128,7 +128,8 @@ export namespace ResolvedHeaderedItemConfig {
     readonly content: ResolvedComponentItemConfig[];
-    readonly activeItemIndex: number;
+    /** The index of the active item in the Stack.  Only undefined if the Stack is empty. */
+    readonly activeItemIndex: number|undefined;
```
whitespace around the `|`

martin31821 commented on src/ts/controls/header.ts:
```diff
@@ -499,7 +498,7 @@ export class Header extends EventEmitter {
 export namespace Header {
     /** @internal */
-    export type GetActiveComponentItemEvent = (this: void) => ComponentItem;
+    export type GetActiveComponentItemEvent = (this: void) => ComponentItem|undefined;
```
once again whitespace

martin31821 commented on src/ts/controls/tabs-container.ts:
```diff
@@ -128,7 +128,7 @@ export class TabsContainer {
      * Pushes the tabs to the tab dropdown if the available space is not sufficient
      */
-    updateTabSizes(availableWidth: number, activeComponentItem: ComponentItem): void {
+    updateTabSizes(availableWidth: number, activeComponentItem: ComponentItem|undefined): void {
```
and one more whitespace

martin31821 commented on src/ts/controls/tabs-container.ts:
```diff
@@ -143,8 +143,12 @@ export class TabsContainer {
     }
 
-    tryUpdateTabSizes(dropdownActive: boolean, availableWidth: number, activeComponentItem: ComponentItem): boolean {
+    tryUpdateTabSizes(dropdownActive: boolean, availableWidth: number, activeComponentItem: ComponentItem|undefined): boolean {
```
whitespace

martin31821 commented on src/ts/items/stack.ts:
```diff
@@ -30,7 +30,7 @@ export class Stack extends ComponentParentableItem {
     private readonly _maximisedEnabled: boolean;
     /** @internal */
-    private _activeComponentItem: ComponentItem;
+    private _activeComponentItem: ComponentItem|undefined;
```
whitespace

martin31821 commented on src/ts/items/stack.ts:
```diff
@@ -245,20 +245,18 @@ export class Stack extends ComponentParentableItem {
     }
 
-    getActiveComponentItem(): ComponentItem {
+    getActiveComponentItem(): ComponentItem|undefined {
```
whitespace

martin31821 commented on src/ts/items/stack.ts:
```diff
@@ -430,9 +428,15 @@ export class Stack extends ComponentParentableItem {
-        const activeItemIndex = this.contentItems.indexOf(this._activeComponentItem);
-        if (activeItemIndex < 0) {
-            throw new AssertError('STC69221');
+        let activeItemIndex: number|undefined;
```
whitespace

martin31821 commented on src/ts/items/stack.ts:
```diff
@@ -245,20 +245,18 @@ export class Stack extends ComponentParentableItem {
-            result = null;
-        }
+        const activeComponentItem = this.getActiveComponentItem();
+        result = activeComponentItem === undefined ? null : activeComponentItem;        
```
could be shortened to:
```
return this.getActiveComponentItem() ?? null;
```

martin31821 reviewed (changes requested):
Looks good apart from the few points I've annotated.

Andrew-Bx commented:
Thanks @martin31821; I've addressed your comments.

Aside 1: it's a shame the whitespace issue isn't flagged by the linter (seems the space-infix-ops rule flags whitespace around the | operator when it's between values, but not between types -- I think this is https://github.com/typescript-eslint/typescript-eslint/issues/51 )

Aside 2: I realised I hadn't been running the linter, so I did.. and then fixed those issues too.

martin31821 reviewed (approved):
LGTM

Andrew-Bx commented:
(just rebased on latest master)
